File: src/main/kotlin/dev/slne/surf/essentialsold/utils/arguments/GameModeArgument.java
package dev.slne.surf.essentialsold.utils.arguments;

import dev.jorel.commandapi.arguments.ArgumentSuggestions;
import dev.jorel.commandapi.arguments.CustomArgument;
import dev.jorel.commandapi.arguments.StringArgument;
import dev.slne.surf.essentialsold.utils.brigadier.Exceptions;
import org.bukkit.GameMode;

import java.util.Arrays;
import java.util.Locale;
import java.util.concurrent.CompletableFuture;

/**
 * Custom argument for {@link GameMode}
 */
public class GameModeArgument extends CustomArgument<GameMode, String> {
    /**
     * A {@link GameMode} argument
     *
     * @param nodeName the name of the node for this argument
     */
    public GameModeArgument(String nodeName) {
        super(new StringArgument(nodeName), info -> {
            try {
                return GameMode.valueOf(info.currentInput().toUpperCase(Locale.ENGLISH)); // try to parse the input as a GameMode
            } catch (IllegalArgumentException e) {
                throw CustomArgumentException.fromAdventureComponent(Exceptions.ERROR_INVALID_GAME_MODE.message(info.currentInput())); // throw an error if the input is not a GameMode
            }
        });
        replaceSuggestions(ArgumentSuggestions.stringCollectionAsync(info -> CompletableFuture.supplyAsync(() -> Arrays.stream(GameMode.values()) // suggest all GameModes
                .map(gameMode -> gameMode.name().toLowerCase(Locale.ENGLISH))
                .toList())));
    }
}

File: src/main/kotlin/dev/slne/surf/essentials/command/argument/GameModeArgument.kt
package dev.slne.surf.essentials.command.argument

import dev.jorel.commandapi.CommandTree
import dev.jorel.commandapi.arguments.Argument
import dev.jorel.commandapi.arguments.ArgumentSuggestions
import dev.jorel.commandapi.arguments.CustomArgument
import dev.jorel.commandapi.arguments.StringArgument
import dev.slne.surf.api.core.messages.adventure.buildText
import dev.slne.surf.essentials.util.permission.EssentialsPermissionRegistry
import org.bukkit.GameMode
import org.bukkit.permissions.Permissible

class GameModeArgument(nodeName: String) :
    CustomArgument<GameMode, String>(StringArgument(nodeName), { info ->
        val input = info.input.lowercase()
        val gameMode = getGameMode(input)
            ?: throw CustomArgumentException.fromAdventureComponent {
                buildText {
                    appendErrorPrefix()
                    error("Der Spielmodus wurde nicht gefunden.")
                }
            }

        val sender = info.sender
        if (sender != null) {
            if (!hasPermissionForGameMode(sender, gameMode)) {
                throw CustomArgumentException.fromAdventureComponent {
                    buildText {
                        appendErrorPrefix()
                        error("Du hast keine Berechtigung für diesen Spielmodus.")
                    }
                }
            }
        }

        gameMode
    }) {

    init {
        replaceSuggestions(
            ArgumentSuggestions.stringCollection { sender ->
                GameMode.entries
                    .filter { hasPermissionForGameMode(sender.sender, it) }
                    .map { it.name.lowercase() }
            }
        )
    }
}

fun hasPermissionForGameMode(permissible: Permissible, gameMode: GameMode) =
    permissible.hasPermission(EssentialsPermissionRegistry.GAME_MODE_WILDCARD) ||
            permissible.hasPermission(permissionForGameMode(gameMode))

fun permissionForGameMode(gameMode: GameMode) =
    "${EssentialsPermissionRegistry.GAME_MODE_BASE}.${gameMode.name.lowercase()}"

private fun getGameMode(gameModeValue: String) = when (gameModeValue) {
    "survival", "s", "0" -> GameMode.SURVIVAL
    "creative", "c", "1" -> GameMode.CREATIVE
    "adventure", "a", "2" -> GameMode.ADVENTURE
    "spectator", "sp", "3" -> GameMode.SPECTATOR
    else -> null
}

inline fun Argument<*>.gameModeArgument(
    nodeName: String,
    optional: Boolean = false,
    block: Argument<*>.() -> Unit = {}
): Argument<*> = then(
    GameModeArgument(nodeName).setOptional(optional).apply(block)
)

inline fun CommandTree.gameModeArgument(
    nodeName: String,
    optional: Boolean = false,
    block: Argument<*>.() -> Unit = {}
): CommandTree = then(
    GameModeArgument(nodeName).setOptional(optional).apply(block)
)
